Kotlin: Support Kotlin 2.3.0

andersfugmann opened the pull request:
This PR add support for Kotlin 2.3.0

To support Kotlin 2.3.0-Beta2, and update of the `rules_kotlin` bazel package needs to be updated to 2.2.0. This change drops  support for Kotlin 1.6.x and 1.7.x.

igfoo commented on java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt:
```diff
@@ -2973,12 +2975,22 @@ open class KotlinFileExtractor(
+                    // For Kotlin < 2.3, s.delegate is not-nullable. Cast to a be nullable,
+                    // as a workaround to silence warnings for kotlin < 2.3 about the elvis
+                    // operator being redundant.
+                    // For Kotlin >= 2.3, the cast is redundant, so we need to silence that warning
```
Do you know why it is now nullable? Is that actually an error?

hfhbd commented:
BTW Kotlin 2.3.0 is released, so you can update 2.3.0-Beta2 to 2.3.0

sgammon commented:
@andersfugmann we are blocked on this, can i help in any way? i tried to complete it on our own fork but couldn't, for version `2.3.0`

andersfugmann commented:
Thanks for the offer to help @sgammon. We have been making progress and expect to have a working solution this week.  Support for Kotlin 2.3 is expected to be available in CodeQL 2.24.1 (Early February)

andersfugmann commented on java/kotlin-extractor/src/main/kotlin/utils/versions/v_2_3_0-Beta2/getJvmModuleNameForDeserializedDescriptor.kt:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    return null
+}
```
Good catch. I'll remove

andersfugmann commented:
@igfoo, I'm marking as ready for review. 
* Addressed your review comments
* Removed support for Kotlin 1.6 and 1.7, and removed `versions` files for these versions and cleaned up the versions files (Removing redundant files and `applied`/inlined files where there were only one file present under `versions`
* I've accepted the test changes, which have been mainly column changes, apart from super constructors - see this [comment](https://github.com/github/codeql/pull/20965#issuecomment-3761094028).
* Updated Kotlin assets to 2.3.0

igfoo commented on java/kotlin-extractor/src/main/kotlin/utils/versions/rollup.sh:
```diff
@@ -0,0 +1,20 @@
+        cp $f ../$DEST
+    fi
+    echo ": $f"
+done
```
I assume that this is for dropping support for old Kotlin versions, in which case I think we ought to document it somewhere, or it will be forgotten about next time. (the best place is currently not in this repo, though, so that won't be part of this PR)

andersfugmann commented on java/ql/integration-tests/kotlin/all-platforms/gradle_kotlinx_serialization/PrintAst.expected:
```diff
@@ -21,35 +21,6 @@ app/src/main/kotlin/testProject/App.kt:
 #    0|         3: [Parameter] serializationConstructorMarker
 #    0|           0: [TypeAccess] SerializationConstructorMarker
-#    7|       5: [BlockStmt] { ... }
-#    7|         0: [ExprStmt] <Expr>;
```
I've merged https://github.com/github/codeql/pull/21216, and will rebase to understand better what the test changes might be. I do expect some changes as we are using an newer serialization plugin, but I expect way fewer changes to the AST.
